File: cpp/leetcode/editor/cn/[977]有序数组的平方.cpp
#include <vector>
#include "util.h"
#include <map>
#include <string>
#include <utility>
#include <memory>

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
public:
    vector<int> sortedSquares(vector<int> &nums) {
        vector<int> r(nums.size());

        auto left = nums.begin();
        auto right = nums.end() - 1;
        auto p = r.rbegin();
        while (distance(left, right) >= 0) {
            if (abs(*left) > abs(*right)) {
                *p = *left * *left;
                left++;
            } else {
                *p = *right * *right;
                right--;
            }
            p++;
        }
        return r;
    }
};
//leetcode submit region end(Prohibit modification and deletion)


int main(int argv, char *argc[]) {
    Solution s;
    vector<int> v = {-4, -1, 0, 3, 10};
    s.sortedSquares(v);
}

File: scala/leetcode/src/main/scala/leetcode/editor/cn/[977]有序数组的平方.scala
package leetcode.editor.cn
object SquaresOfASortedArray{
//leetcode submit region begin(Prohibit modification and deletion)
object Solution {
    def sortedSquares(nums: Array[Int]): Array[Int] = {
      val res = Array.ofDim[Int](nums.length)
      var left = 0
      var right = nums.length - 1
      var p = nums.length - 1

      while left <= right do
        if math.abs(nums(left)) > math.abs(nums(right)) then
          res(p) = nums(left) * nums(left)
          left += 1
        else
          res(p)= nums(right) * nums(right)
          right -= 1
        p -= 1

      res
    }
}
//leetcode submit region end(Prohibit modification and deletion)

}
